### PR TITLE
[TextField / Labelled] Add character count

### LINKED ISSUE #491: TextField remaining characters count
# Feature request summary

It would be nice if TextFields had the option to show the number of characters remaining when `maxLength` prop is set (see example below), or just the number of characters in the input if no `maxLength` is set.

screen shot 2018-10-29 at 1 48 42 pm

Perhaps a simple `showCharacterCount` prop can be used to enable/disable the character count being displayed.

## Rationale

It is quite common when building forms to have text input fields that are character limited. Simply setting a max length on the field means that the user has no visual feedback that they have run out of characters except for that when they continue to type characters are no longer added to the input.

This also means that users can't see how many characters they have remaining while trying to decide what to type.

---

## Possible Implementation
Because the `TextField` component uses the `Labelled` component to show the label and `action` (if it exists), my initial thought was to extend the `Labelled` component to optionally display the number of characters remaining in the `TextField`. But this does mean that `TextField` has to pass optional information about the character count (as well as whether or not to show the count) to `Labelled`.

Another implementation could be simply creating another component that wraps the existing `TextField` and displays the character count. I feel like this might need some interesting css though to get the count to line up properly and adjust its position for when a `label action` is also used (like below).

screen shot 2018-10-29 at 3 46 39 pm

### PR DESCRIPTION
**Pull request by jfangrad**
### WHY are these changes introduced?

Resolves #491

TextField lacks the ability to display the remaining characters available when the TextField has a max character length. This PR adds that ability.

### WHAT is this pull request doing?

Adds in the prop `showCharacterCount` to TextField to enable the displaying of the remaining character count.

The `Labelled` component will then display the number of characters out of the total allowed characters that have been inputted into the TextField.

screen shot 2018-12-03 at 4 21 40 pm

One item of note is that things can get a little busy as can be seen below:

screen shot 2018-11-29 at 4 20 26 pm

@dpersing was giving insight into this on the issue.

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page, TextField} from '@shopify/polaris';

interface State {
  value: string;
}

export default class Playground extends React.Component<{}, State> {
  state = {
    value: '',
  };

  onChange = (value: string) => {
    this.setState({value});
  };

  render() {
    const {value} = this.state;
    return (
      <Page
        title="Playground"
        primaryAction={{content: 'View Examples', url: '/examples'}}
      >
        <TextField
          label="This is where the label lives"
          value={value}
          onChange={this.onChange}
          maxLength={10}
          showCharacterCount
        />
        <TextField
          label="This is where the label lives"
          value={value}
          onChange={this.onChange}
          maxLength={10}
          multiline         
          showCharacterCount
        />
      </Page>
    );
  }
}
```

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated [UNRELEASED.md](https://github.com/Shopify/polaris-react/blob/master/UNRELEASED.md)

### CODE REVIEW

File: src/components/Labelled/Labelled.tsx [hunk 1]
```diff
@@ -48,6 +57,17 @@ export default function Labelled({
     </div>
   ) : null;
 
+  const characterCountMarkup = showCharacterCount &&
+    maxLength && (
+      <span
+        id={`${id}-labeled-character-counter`}
+        className={styles.CharacterCount}
+        aria-live="polite"
```
Comment (dpersing) [inline src/components/Labelled/Labelled.tsx]: This looks great, @jfangrad!

I tested with a few common screen reader/browser combos and the character count help text is read when the field gets focus, and the updates get read. One small recommendation I'd make is to add `aria-atomic="true"` to the `aria-live` container here. Right now, NVDA with Firefox is only reading the updated character count, not the whole string, so the context is not quite as clear. Adding that attribute will cause the whole `<span>` to be read when it changes.

File: src/components/Labelled/Labelled.tsx [hunk 2]
```diff
@@ -36,6 +42,9 @@ export default function Labelled({
   helpText,
   children,
   labelHidden,
+  characterCount,
+  maxLength,
+  showCharacterCount,
```
Comment (danrosenthal) [inline src/components/Labelled/Labelled.tsx]: The `showCharacterCount` prop here feels unnecessary. As an implementor I would expect that if I pass a character count, it should show a character count. And if I don't pass a character count, it should not show a character count.

File: src/components/Labelled/Labelled.tsx [hunk 3]
```diff
@@ -48,6 +57,17 @@ export default function Labelled({
     </div>
   ) : null;
 
+  const characterCountMarkup = showCharacterCount &&
+    maxLength && (
```
Comment (danrosenthal) [inline src/components/Labelled/Labelled.tsx]: Would there ever be a case where this might be used without a maxLength? Does it make sense to only render it if the `maxLength` is provided?

File: src/components/Labelled/Labelled.tsx [hunk 4]
```diff
@@ -48,6 +57,17 @@ export default function Labelled({
...
+      <span
+        id={`${id}-labeled-character-counter`}
+        className={styles.CharacterCount}
+        aria-live="polite"
+      >
+        {`${characterCount} of ${maxLength} characters used`}
```
Comment (danrosenthal) [inline src/components/Labelled/Labelled.tsx]: This string will need to be externalized and internationalized. Some languages will change the position of words here, and we avoid hard-coding strings in our components. This is done using the `AppProvider` component and the `intl` context it provides. Strings live in the `en.json` file at the root of this project.

Grepping around the project should show you some examples of how that is done.

File: src/components/Labelled/Labelled.tsx (hunk 1 above)
Comment (danrosenthal) [reply src/components/Labelled/Labelled.tsx]: Bumping @dpersing's suggestion to add `aria-atomic="true"` to the `aria-live` container.

File: src/components/TextField/tests/TextField.test.tsx [hunk 5]
```diff
@@ -595,13 +595,18 @@ describe('<TextField />', () => {
...
         />,
       );
 
       expect(textField.find(Labelled)).toHaveLength(1);
       expect(textField.find(Labelled).prop('label')).toBe('TextField');
       expect(textField.find(Labelled).prop('id')).toBe('MyField');
       expect(textField.find(Labelled).prop('helpText')).toBe('Help text');
+      expect(textField.find(Labelled).prop('showCharacterCount')).toBe(true);
+      expect(textField.find(Labelled).prop('characterCount')).toBe(0);
+      expect(textField.find(Labelled).prop('maxLength')).toBe(5);
```
Comment (danrosenthal) [inline src/components/TextField/tests/TextField.test.tsx]: This could also use a test for when the field's value is of length greater than zero, asserting that the character count does in fact count the characters.

**Comment by solonaarmstrong** [comment]
I wonder if it would make more sense to position the text count here: 
screen shot 2018-12-14 at 7 03 08 am

This is how it is already implemented across the admin. It looks like it's using text style as a sibling to the label.

**Comment by danrosenthal** [comment]
I'm also in favor of that positioning, @solonaarmstrong. cc: @sarahill to confirm

**Comment by jfangrad** [comment]
@solonaarmstrong @danrosenthal I also agree and think that it looks better in that position. The one reason, that came up in the issue, for putting it below was that if the textfield had a lot of text in it the user might not be able to see the count as they type if it was at the top of the field.

Its probably a rare case but just thought I would bring that up again to be included in the consideration.

**Comment by danrosenthal** [comment]
> The one reason, that came up in the issue, for putting it below was that if the textfield had a lot of text in it the user might not be able to see the count as they type if it was at the top of the field.

I think that's an acceptable limitation.

**Comment by sarahill** [comment]
I like @solonaarmstrong suggestion for placement.

Couple questions/thoughts:
- How would this look on smaller screens.
    - We could either not show the count on or reduce it to just the count `4 / 70`.
- Do we need to worry about a long label that could potentially run into the character count?
    - For this we could set a max-width and allow the label to wrap if necessary and base align the label + character count.

**Comment by solonaarmstrong** [comment]
@sarahill This is how it currently behaves on small screens in the admin (for SEO, as an example). It simply needs to stack.

screen shot 2018-12-14 at 9 07 37 am

File: src/components/Labelled/Labelled.tsx (hunk 3 above)
Comment (solonaarmstrong) [reply src/components/Labelled/Labelled.tsx]: I had the same feeling. If there is no `maxLength`, I would change the copy to `n characters used` or simply `n characters`.

File: src/components/Labelled/Labelled.tsx [hunk 6]
```diff
@@ -26,6 +26,12 @@ export interface Props {
   children?: React.ReactNode;
   /** Visually hide the label */
   labelHidden?: boolean;
+  /** Character count in the labled input */
+  characterCount?: number;
+  /** Maximum character length for an input */
+  maxLength?: number;
```
Comment (solonaarmstrong) [inline src/components/Labelled/Labelled.tsx]: It seems a bit unclear what has a `maxLength`: the input or the label. This might want to be named `inputMaxLength`.

File: src/components/TextField/TextField.tsx [hunk 7]
```diff
@@ -96,6 +96,8 @@ export interface BaseProps {
   ariaActiveDescendant?: string;
   /** Indicates what kind of user input completion suggestions are provided */
   ariaAutocomplete?: string;
+  /** Indicates whether the character count should be displayed or not */
```
Comment (solonaarmstrong) [inline src/components/TextField/TextField.tsx]:
```suggestion
  /** Indicates whether or not the character count should be displayed */
```

File: src/components/Labelled/Labelled.tsx [hunk 8]
```diff
@@ -26,6 +26,12 @@ export interface Props {
   children?: React.ReactNode;
   /** Visually hide the label */
   labelHidden?: boolean;
+  /** Character count in the labled input */
```
Comment (solonaarmstrong) [inline src/components/Labelled/Labelled.tsx]:
```suggestion
  /** Character count in the labeled input */
```

Comment (solonaarmstrong) [reply src/components/Labelled/Labelled.tsx]: Or better yet
```suggestion
  /** Displays the number of characters entered in the input */
```

**Comment by jfangrad** [comment]
@danrosenthal @solonaarmstrong I made the changes to the placement etc.

One thing that I ran into is that the placement of the label action gets a little weird when also showing the character count. Currently this is what I have it doing when both are present:
screen shot 2018-12-17 at 1 34 08 pm

Its probably a super rare case to have all three...
To have the label action always be on the far right I would have to move the character count code down another level into the `Label` component which wouldn't be super nice.

label action will still be on the far right if the `showCharacterCount` is false

**Comment by jfangrad** [comment]
@danrosenthal I definitely agree with you. It felt weird putting all the logic in Labelled.

I like the idea of passing it through the action prop, my only concern is that you will then not be able to have both at the same time. I know for my project that isn't a concern. I guess if it becomes a problem in the future that can be addressed then?

My main question then is if someone passes a `labelAction` prop and the `showCharacterCount` prop to the `TextField` which should we pass to the `Labelled` components `action` prop? Which takes precedence? My initial thought is action should probably take precedence since it has been around longer.

I am also worried about people trying to do an action and a character count and wondering why one of them doesn't show up (But maybe just saying why in the docs is enough?)

**Comment by danrosenthal** [comment]
It feels to me like there should only be one element ever rendered in that slot, similar to how we handle it in [Polaris rails](https://polaris-rails.shopify.com/components/text_field#side-controls) with side controls.

This feels like a good time to bring in another opinion, so I've asked @tmlayton to weigh in.

**Comment by tmlayton** [comment]
A couple things off the bat:

- There are visual regressions here where labels are being rendered when they shouldn’t (check the Percy link)
- Let’s add an example for this

--- 

As to the approach, yes this is a concern of the text field, at least the logic and string generation. The label should be naive here and only care about layout and styling of the string, not its contents. It seems to me we are adding two features, 1) text field generating a character count / remaining characters string and 2) labelled supporting upper right aligned tertiary text.

Similar to how we pass help text to labelled, I would expect a prop like `tertiaryText` or `infoText` on labelled and we simply pass the generated count string from text field to the new labelled prop. I would also expect `labelHidden` to not visually render this `infoText`, but still set semantic/aria attributes.

**Comment by tmlayton** [comment]
A new prop in addition to label action – I think using the action prop distorts the API.

> Also since the labelHidden prop doesn't hide the labelAction why should it hide the character count (or infoText).

I can see the reasoning here, but actions need to interacted with thus have a dependency on being rendered to serve any purpose. Labels, however, don't need to be rendered and can still serve a purpose as aria labels, etc. So my idealism-based position would be if it is purely information (nothing to interact with) `labelHidden` should hide it. But...

> I can think of a couple cases in a project I am working on where we want a hidden label but would still want the character count to show.

...I think pragmatism will be better here 😄 (aka `labelHidden` hides the label, not the character count)

**Comment by sarahill** [comment]
Some quick suggestions for displaying the character count. I think this would work with the `labelHidden`. I don't know if we want to change it that much but just some thoughts.

image

cc: @dpersing

**Comment by dpersing** [comment]
I like including the count as part of the field! The only case where things might be weird is if the field is very long and magnification cuts off the number.

For non-visual users, I think we'd want to provide a bit more context for the number. Something like "X characters remaining" would be good (and still with the `aria-atomic="true"` for consistency).

**Comment by jfangrad** [comment]
@danrosenthal how do you feel about moving to the design Sara has suggested? It does eliminate a lot of the issues we were running into and make the implementation potentially a lot simpler.

**Comment by danrosenthal** [comment]
@jfangrad I feel good about it. Would you mind taking that on?

File: src/components/TextField/TextField.scss [hunk 9]
```diff
@@ -170,6 +170,22 @@ $stacking-order: (
...
+  z-index: z-index(contents, $stacking-order);
+  margin-right: $backdrop-horizontal-spacing;
+  pointer-events: none;
+}
+
+.MaxCharacters {
+  color: color('red');
+}
+
+.FlexEnd {
```
Comment (danrosenthal) [inline src/components/TextField/TextField.scss]: I think this could use a more semantic name that describes its purpose.

File: src/components/TextField/TextField.scss [hunk 10]
```diff
@@ -170,6 +170,22 @@ $stacking-order: (
...
+  align-self: flex-end;
+  margin-bottom: rem(4px);
```
Comment (danrosenthal) [inline src/components/TextField/TextField.scss]: What's the purpose of setting a `margin-bottom` here? Also, this could use `spacing(extra-tight)`.

File: src/components/TextField/TextField.scss [hunk 11]
```diff
@@ -170,6 +170,22 @@ $stacking-order: (
   pointer-events: none;
 }
 
+.CharacterCount {
+  @include text-emphasis-subdued;
+  z-index: z-index(contents, $stacking-order);
```
Comment (danrosenthal) [inline src/components/TextField/TextField.scss]: Is the `z-index` necessary to specify here?

File: src/components/TextField/TextField.tsx [hunk 12]
```diff
@@ -204,6 +211,32 @@ export default class TextField extends React.PureComponent<Props, State> {
...
+      maxLength
+        ? 'Polaris.TextField.characterCountWithMaxLength'
+        : 'Polaris.TextField.characterCount',
+      {count: characterCount},
+    );
+
+    const characterCountClassName = classNames(
+      styles.CharacterCount,
+      multiline && styles.FlexEnd,
+      maxLength && characterCount === 0 && styles.MaxCharacters,
```
Comment (danrosenthal) [inline src/components/TextField/TextField.tsx]: If possible, it would be great to add code coverage for these conditions as well. (224 and 225)

**Comment by danrosenthal** [comment]
@dpersing you can use the deployed example to tophat here: https://polaris-react-pr-709.herokuapp.com/text-field/text-field-with-character-count

**Comment by dpersing** [comment]
Tested the updated version with VoiceOver/Safari on macOS and iOS and NVDA with Firefox on Windows 10.

On macOS, VoiceOver is totally fine with the `aria-label` for the live region update. On iOS, and on Windows, the `aria-label` is skipped, but the region content is read out as expected. Since a user will hear the initial "X characters remaining" information when the field gets focus (or after a short delay), I think this is good to ship with.

**Comment by solonaarmstrong** [comment]
Personally, I would do this a bit differently. I would either offer the consumer the option to count up or count down. Or I would be opinionated about counting up.

In production, the character count reads as `10 out of 30 characters used` (it counts up). I think counting up is more common in the wild. I initially was confused why 11 characters were displaying `9`. There is also no way to append `characters used` or `characters left`.

**Comment by solonaarmstrong** [comment]
As well, @jfangrad if you want me to write some more tests, I like writing tests. Just don't want to step on your toes.

File: src/components/TextField/TextField.tsx [hunk 13]
```diff
@@ -204,6 +211,32 @@ export default class TextField extends React.PureComponent<Props, State> {
...
+    );
+
+    const characterCountMarkup = showCharacterCount ? (
+      <div
+        id={`${id}-character-counter`}
```
Comment (solonaarmstrong) [inline src/components/TextField/TextField.tsx]:
```suggestion
         id={`${id}CharacterCounter`}
```

File: src/components/TextField/tests/TextField.test.tsx [hunk 14]
```diff
@@ -305,6 +305,41 @@ describe('<TextField />', () => {
...
+        <TextField
+          value="test"
+          showCharacterCount
+          label="TextField"
+          id="MyField"
+          onChange={noop}
+        />,
+      );
+
+      const characterCount = textField.find('#MyField-character-counter');
```
Comment (solonaarmstrong) [inline src/components/TextField/tests/TextField.test.tsx]: This is probably a good case for using `findByTestID`. Throw a `testID` on the div element and
```suggestion
       const characterCount = findByTestID(textField, 'CharacterCounter');
```

File: src/components/TextField/tests/TextField.test.tsx [hunk 15]
```diff
@@ -305,6 +305,41 @@ describe('<TextField />', () => {
...
+
+      expect(characterCount.text()).toBe('4');
+    });
+
+    it('counts characters in input field with max length', () => {
```
Comment (solonaarmstrong) [inline src/components/TextField/tests/TextField.test.tsx]:
```suggestion
     it('counts remaining characters in input field with maxLength', () => {
```

File: src/components/TextField/tests/TextField.test.tsx [hunk 16]
```diff
@@ -305,6 +305,41 @@ describe('<TextField />', () => {
     });
   });
 
+  describe('characterCount', () => {
+    it('counts characters in input field', () => {
```
Comment (solonaarmstrong) [inline src/components/TextField/tests/TextField.test.tsx]:
```suggestion
     it('displays number of characters entered in input field', () => {
```

File: src/components/TextField/README.md [hunk 17]
```diff
@@ -737,6 +737,36 @@ Use to show that a textfield is not available for interaction. Most often used i
 <TextField label="Store name" disabled />
 ```
 
+### Text field with character count
+
+<!-- example-for: web -->
+
+Use to show the number of characters remaining in a text field with a max length. If max length is not provided it will display the current number of characters in the text field.
```
Comment (solonaarmstrong) [inline src/components/TextField/README.md]: Same comment as I've already left, but I find this a bit unclear.

File: src/components/TextField/TextField.scss (hunk 11 above)
Comment (TheEnzoMar) [reply src/components/TextField/TextField.scss]: This is needed because of https://github.com/Shopify/polaris-react/blob/master/src/components/TextField/TextField.scss#L140 which sets a `z-index`

**Comment by TheEnzoMar** [comment]
Hi @solonaarmstrong and @danrosenthal 
Justin's won't be able to finish the PR as his internship ended, so I'm going to taking over it. I'm currently reviewing the code and your comments/feedback. Cheers!

**Comment by TheEnzoMar** [comment]
Hi @ry5n, I tested what the text field w. character count looks like when other attributes are enabled. Here are the results.

image

Let me know what you think and/or any comments or concerns.

Cheers!

**Comment by TheEnzoMar** [comment]
@solonaarmstrong, regarding
> Personally, I would do this a bit differently. I would either offer the consumer the option to count up or count down - with clear indications of what the number represents. Or I would be opinionated about counting up...

I see value in your comment and agree that a more consistent way of counting would be better. I think for simplicity we should count up, and when there is a max character limit the label should denote `x of y`.

What do you think of something like,

image

**Comment by solonaarmstrong** [comment]
@TheEnzoMar Yes, that makes much more sense to me.

File: src/components/TextField/README.md (hunk 17 above)
Comment (tmlayton) [reply src/components/TextField/README.md]:
```suggestion
Use to show the number of characters remaining in a text field with a max length. If max length is not provided it will display the current number of characters in the text field.
```

**Comment by dpersing** [comment]
Hey @TheEnzoMar! Sounds like there's agreement on the counting method. Let us know if we can provide any additional help on this PR.

**Comment by emarchak** [comment]
It ended up making more sense to provide a fixed line-height to the CharacterCount instead of shifting around the margins if it's a multiline text field. I believe we still need the `.AlignFieldBottom` since the layout had inconsistent behaviour when I applied `align-self: flex-end;` to a single line input field. 

image

I believe this should be good to go pending the two ✅ and someone approving percy. @solonaarmstrong and @dpersing could you please take another look?

**Comment by dpersing** [comment]
Hi @emarchak! It looks like the `aria-label` for the "X of Y" content is reading out "X characters remaining", which is the opposite of the visual. For example, when 0 characters have been typed, it says "0 characters remaining", whereas really it should be "10 characters remaining" since none have been typed yet:

```
<div id="TextField1CharacterCounter" class="TextField-CharacterCount_7V_ks" aria-label="0 characters remaining" aria-live="polite" aria-atomic="true">0/10</div>
```

I'm thinking that "X of Y characters used" would make more sense with the visual?

**Comment by emarchak** [comment]
Fixed!

File: src/locales/en.json [hunk 18]
```diff
@@ -192,6 +192,11 @@
 
     "Tag": {
       "ariaLabel": "Remove {children}"
+    },
+
+    "TextField": {
+      "characterCount": "{count} characters",
+      "characterCountWithMaxLength": "{count} of {limit} characters used"
```
Comment (dpersing) [inline src/locales/en.json]: 🎉

**Comment by fchienvuhoang** [comment]
I'm using bellow code example but not work, my Polaris v3.8.0, any update?
```
 <TextField
        label="Store name"
        value={this.state.value}
        onChange={this.handleChange}
        maxLength={20}
        showCharacterCount
        multiline
      />
```

**Comment by solonaarmstrong** [comment]
@fchienvuhoang I just tried and I see it in there. 

screen shot 2019-02-22 at 9 31 15 am
